### ISSUE TITLE
[1.4.0-RC1] Home and Delete keys do not work on Windows 10

### ISSUE THREAD

**Issue by melezov** [sbt/sbt#5828]
## steps

On Windows 10 x64, JDK 11/14/...
Run sbt (using the launcher directly to skip sbtconfig.txt / sbtopts):

```
   r:\test>java -Duser.home=r:\empty -jar sbt-launch-1.4.0-RC1.jar
   [info] welcome to sbt 1.4.0-RC1 (AdoptOpenJDK Java 14.0.1)
   [info] loading project definition from R:\test\project
   [info] set current project to test (in build file:/R:/test/)
```

Type "special" keys such as Home/End/Delete.

## problem 

The "special" keys such as Home/End/Delete produce garbage (raw control sequence) on the command line.

```
   r:\test>java -Duser.home=r:\empty -jar sbt-launch-1.4.0-RC1.jar
   [info] welcome to sbt 1.4.0-RC1 (AdoptOpenJDK Java 14.0.1)
   [info] loading project definition from R:\test\project
   [info] set current project to test (in build file:/R:/test/)
   [info]
   [info] Here are some highlights of this release:
   [info]   - Build server protocol (BSP) support
   [info]   - sbtn: a native thin client for sbt
   [info]   - VirtualFile + RemoteCache: caches build artifacts across different machines
   [info]   - ThisBuild / versionScheme to take the guessing out of eviction warning
   [info] See http://eed3si9n.com/sbt-1.4.0-beta for full release notes.
   [info] Hide the banner for this release by running `skipBanner`.
   [info] sbt server started at local:sbt-server-83ade0559bc655750a72
   sbt:test> [3~[3~[3~skipBanner[1~[4~[1~[1~                        
```

## expectation

Home is interpreted.

## notes

This is a regression from the same behavior in SBT 1.x which didn't work before, but got fixed.
I'm assuming it has to do with new JLine.
Luckily, the arrow keys seem to work, so you can workaround these.

I've tried to run it with `-Djline.terminal=jline.UnsupportedTerminal` and then the keys such as Home/End/Delete seem to work out of the box, but I'm not sure if this disables some other useful functionality.

**Comment by eed3si9n** [comment]
@melezov Thanks for the report!

**Comment by eatkins** [comment]
We might need to do something like what is done here: https://github.com/jline/jline3/issues/396. This one is a bit difficult for me to debug because I don't have a keyboard with those keys on it.

**Comment by melezov** [comment]
If there's any way a Windoze person can help diagnose/debug, just scream at me.

[Edit]: Aha, I see, this is a problem on Nux terminals as well. I see Home/End is freaking out when I SSH into a Centos as well.

**Comment by eatkins** [comment]
I can reproduce forward delete not working on windows at least. JLine is a nightmare to debug so I can't promise a fix though.

**Comment by eatkins** [comment]
Yeah. I'm sorry but I am not going to fix this. The internal structure of JLine is far too difficult for me to figure out what is going on without serious surgery. Feel free to open an issue with https://github.com/jline/jline3. Relevant code is in this file: https://github.com/sbt/sbt/blob/develop/internal/util-logging/src/main/scala/sbt/internal/util/JLine3.scala.

**Comment by eed3si9n** [comment]
I'm guessing that this comes down to these files:
- https://github.com/jline/jline3/blob/jline-parent-3.16.0/terminal/src/main/resources/org/jline/utils/screen-256color.caps
- https://github.com/jline/jline3/blob/jline-parent-3.16.0/terminal/src/main/resources/org/jline/utils/windows.caps

What terminal program are you using? [Windows Terminal](https://github.com/microsoft/terminal), Windows Console, or something else? Microsoft has been revamping the terminal so it can run Ubuntu and all the ncurses programs, so I wouldn't be surprised that control sequence for Home has also changed for Ubuntu compatibility. Also do you (or your terminal) set `TERM` environment variable?

**Comment by eatkins** [comment]
I should add that right now the capabilities are hardcoded to ansi for all platforms: https://github.com/sbt/sbt/blob/8160035f41b3e8dd1d6b8b7b884ba91391c6623a/internal/util-logging/src/main/scala/sbt/internal/util/JLine3.scala#L45-L54 https://github.com/sbt/sbt/blob/8160035f41b3e8dd1d6b8b7b884ba91391c6623a/internal/util-logging/src/main/scala/sbt/internal/util/JLine3.scala#L83
I tried changing the capabilities to "windows" on my vm and it actually made things worse. For example, the left and right arrows stopped working and forward delete still was still being interpreted as `\e[3~`. I was testing with powershell 7.

**Comment by melezov** [comment]
The behaviour is identical no matter which shell I try out:
Windows Console ("cmd"), Powershell ("ps"), Terminal + PS/cmd, even Cygwin64 terminal

TERM envvar is only set in Cygwin64 to "xterm".

I understand if this goes out of scope of SBT-1.4.0, hopefully Jline folks might offer some advice.

**Comment by eatkins** [comment]
What I have noticed is that my windows terminal behaves very similar to my mac terminal, but jline seems to assume that it will behave differently. There is some logic for handling [windows key codes](https://docs.microsoft.com/en-us/windows/win32/inputdev/virtual-key-codes) https://github.com/jline/jline3/blob/6551251c1f987f4f74b303e4685dd5916f18346b/terminal/src/main/java/org/jline/terminal/impl/AbstractWindowsTerminal.java#L308
but in my terminal at least the normal ansi code for forward delete, `\e[3~`, is sent, not VK_DELETE, `0x2e`. I would assume that because the keybinding is set to ansi that things should work, but they don't. I don't know what JLine is doing to override the ansi key mappings. After working on this for a few hours, I just decided it was too niche to pursue further because the lack of forward delete, home and end do not prevent any sbt features from working. It would be great if someone could provide a fix though.
